book-content/chapters/11-annotations-and-assertions.md

```diff
@@ -0,0 +1,1196 @@
+Throughout this book, we've been using relatively simple type annotations. We've had a look at variable annotations, which help TypeScript know what type a variable should be:
+
+```typescript
+let name: string;
+
+name = "Waqas";
+```
+
+We've also seen how to type function parameters and return types:
+
+```typescript
+function greet(name: string): string {
+  return `Hello, ${name}!`;
+}
+```
+
+These annotations are instructions to TypeScript to tell it what type something should be. If we return a `number` from our `greet` function, TypeScript will show an error. We've told TypeScript that we're returning a `string`, so it expects a `string`.
+
+But there are times when we _don't_ want to follow this pattern. Sometimes, we want to let TypeScript figure it out on its own.
+
+And sometimes, we want to lie to TypeScript.
+
+In this chapter, we'll look at more ways to communicate with TypeScript's compiler via annotations and assertions.
+
+## Annotating Variables vs Values
+
+There's a difference in TypeScript between annotating _variables_ and _values_. The way they conflict can be confusing.
+
+### When You Annotate A Variable, The Variable Wins
+
+Let's look again at the variable annotation we've seen throughout this book.
+
+In this example, we're declaring a variable `config` and annotating it as a `Record` with a string key and a `Color` value:
+
+```typescript
+type Color =
+  | string
+  | {
+      r: number;
+      g: number;
+      b: number;
+    };
+
+const config: Record<string, Color> = {
+  foreground: { r: 255, g: 255, b: 255 },
+  background: { r: 0, g: 0, b: 0 },
+  border: "transparent",
+};
+```
+
+Here, we're annotating a variable. We're saying that `config` is a `Record` with a string key and a `Color` value. This is useful, because if we specify a `Color` that doesn't match the type, TypeScript will show an error:
+
+```ts twoslash
+// @errors: 2353
+type Color =
+  | string
+  | {
+      r: number;
+      g: number;
+      b: number;
+    };
+
+// ---cut---
+const config: Record<string, Color> = {
+  border: { incorrect: 0, g: 0, b: 0 },
+};
+```
+
+But there's a problem with this approach. If we try to access any of the keys, TypeScript gets confused:
+
+```ts twoslash
+// @errors: 2339
+type Color =
+  | string
+  | {
+      r: number;
+      g: number;
+      b: number;
+    };
+
+const config: Record<string, Color> = {};
+
+// ---cut---
+config.foreground.r;
+```
+
+Firstly, it doesn't know that foreground is defined on the object. Secondly, it doesn't know whether foreground is the `string` version of the `Color` type or the object version.
+
+This is because we've told TypeScript that `config` is a `Record` with a any number of string keys. We annotated the variable, but the actual _value_ got discarded. This is an important point - when you annotate a variable, TypeScript will:
+
+1. Ensure that the value passed to the variable matches the annotation.
+2. Forget about the value's type.
```

Comment:
not always because there is something that internally is called an assignment reduced type:
```ts
const test: 1 | 2 = 1;
test;
// ^? const test: 1
```